### PR TITLE
[SPARK-44391][SQL] Check the number of argument types in `InvokeLike`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to check the number of argument types in the `InvokeLike` expressions. If the input types are provided, the number of types should be exactly the same as the number of argument expressions.

### Why are the changes needed?
1. This PR checks the contract described in the comment explicitly:
https://github.com/apache/spark/blob/d9248e83bbb3af49333608bebe7149b1aaeca738/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala#L247 

that can prevent the errors of expression implementations, and improve code maintainability.

2. Also it fixes the issue in the `UrlEncode` and `UrlDecode`.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running the related tests:
```
$ build/sbt "test:testOnly *UrlFunctionsSuite"
$ build/sbt "test:testOnly *DataSourceV2FunctionSuite"
```

### CODE REVIEW

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -657,6 +657,11 @@
           "The <exprName> must be between <valueRange> (current value = <currentValue>)."
         ]
       },
+      "WRONG_NUM_ARG_TYPES" : {
+        "message" : [
+          "The expression requires <expectedNum> argument types but the actual number is <actualNum>."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:662]: I feel the number of argument types is a bit confusing to end users. shall we just say number of arguments?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
how far shall we backport it?

File: common/utils/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:662]: If we say that the number of arguments is wrong but actually the user passes correct arguments. That might confuse much more.

**Comment by MaxGekk** [comment]
> how far shall we backport it?

To 3.4, but I expects some conflicts in `branch-3.4`. Let's merge this to master, and I'll open a separate PR with a backport.

**Comment by MaxGekk** [comment]
All GAs passed. Merging to master. Thank you, @cloud-fan for review.

**Pull request closed without merging**